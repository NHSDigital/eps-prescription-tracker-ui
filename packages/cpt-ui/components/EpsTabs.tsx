'use client'
import React from 'react'
import "../assets/styles/tabs.scss"
import { Tabs } from "nhsuk-react-components";
import {
    PRESCRIPTION_SEARCH_TABS
} from "../constants/ui-strings/SearchTabStrings"

export default function EpsTabs() {
    const tabData = PRESCRIPTION_SEARCH_TABS;
    return (
        <Tabs>
            <Tabs.Title>Contents</Tabs.Title>
            <Tabs.List>
                {
                    tabData.map(tabHeader =>
                        <Tabs.ListItem id={tabHeader.targetId} key={tabHeader.title}>{tabHeader.title}</Tabs.ListItem>
                    )
                }
            </Tabs.List>
            {
                tabData.map(tabContent =>
                    <Tabs.Contents id={tabContent.targetId} key={tabContent.title}>
                        <div>{tabContent.title}</div>
                    </Tabs.Contents>
                )
            }
        </Tabs>
    )
}
