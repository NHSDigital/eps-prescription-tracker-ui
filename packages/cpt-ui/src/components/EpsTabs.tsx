import React from "react"
import {Link} from "react-router-dom"
import {Tabs} from "nhsuk-react-components"
import {PRESCRIPTION_SEARCH_TABS} from "@/constants/ui-strings/SearchTabStrings"

interface EpsTabsProps {
  activeTabPath: string;
  children: React.ReactNode;
}

export default function EpsTabs({activeTabPath, children}: EpsTabsProps) {
  return (
    <div className="nhsuk-tabs">
      <Tabs.Title>Contents</Tabs.Title>
      <Tabs.List>
        {PRESCRIPTION_SEARCH_TABS.map((tab) => (
          // we are using a custom list item component here instead of Tabs.ListItem
          // because the built-in functionality of Tabs.ListItem does not allow for
          // the tab to navigate us to separate urls instead it is hardwired to work
          // with target ids, which goes against our intended use case
          <li
            key={tab.link}
            className={`nhsuk-tabs__list-item ${activeTabPath === tab.link ? "nhsuk-tabs__list-item--selected" : ""}`}
            role="presentation"
          >
            <Link
              className="nhsuk-tabs__tab"
              role="tab"
              aria-selected={activeTabPath === tab.link ? "true" : "false"}
              id={`tab_${tab.link.substring(1)}`}
              aria-controls={`panel_${tab.link.substring(1)}`}
              to={tab.link}
            >
              {tab.title}
            </Link>
          </li>
        ))}
      </Tabs.List>
      <div
        className="nhsuk-tabs__panel"
        id={`panel_${activeTabPath.substring(1)}`}
        role="tabpanel"
        aria-labelledby={`tab_${activeTabPath.substring(1)}`}
      >
        {children}
      </div>
    </div>
  )

}
