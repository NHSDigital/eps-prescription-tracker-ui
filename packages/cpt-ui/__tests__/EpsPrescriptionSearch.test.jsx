import React from 'react';
import "@testing-library/jest-dom";
import { fireEvent, render } from "@testing-library/react";
import EpsTabs from "../components/EpsHeader";
import { Tabs } from "nhsuk-react-components";
import {
  PRESCRIPTION_SEARCH_TABS
} from "../constants/ui-strings/SearchTabStrings"

describe('The tabs component', () => {
  it('Matches the snapshot', () => {
    const tabData = PRESCRIPTION_SEARCH_TABS;
    const { container } = render(
      <Tabs>
            <Tabs.Title>Contents</Tabs.Title>
            <Tabs.List>
                {
                    tabData.map(tabHeader =>
                        <Tabs.ListItem id={tabHeader.targetId} key={tabHeader.title}>{tabHeader.title}</Tabs.ListItem>
                    )
                }
            </Tabs.List>
            {
                tabData.map(tabContent =>
                    <Tabs.Contents id={tabContent.targetId} key={tabContent.title}>
                        <div>{tabContent.title}</div>
                    </Tabs.Contents>
                )
            }
        </Tabs>
    );

    expect(container).toMatchSnapshot();
  });

  it('Switches the visibility of tabs when clicked', () => {
    const tabData = PRESCRIPTION_SEARCH_TABS;
    const { container } = render(
      <Tabs>
            <Tabs.Title>Contents</Tabs.Title>
            <Tabs.List>
                {
                    tabData.map(tabHeader =>
                        <Tabs.ListItem id={tabHeader.targetId} key={tabHeader.title}>{tabHeader.title}</Tabs.ListItem>
                    )
                }
            </Tabs.List>
            {
                tabData.map(tabContent =>
                    <Tabs.Contents id={tabContent.targetId} key={tabContent.title}>
                        <div>{tabContent.title}</div>
                    </Tabs.Contents>
                )
            }
        </Tabs>,
    );

    const firstTabLink = container.querySelector('#tab_IdSearch');
    const secondTabLink = container.querySelector('#tab_NhsNumSearch');

    expect(
      firstTabLink?.parentElement?.classList.contains('nhsuk-tabs__list-item--selected'),
    ).toEqual(true);
    expect(
      secondTabLink?.parentElement?.classList.contains('nhsuk-tabs__list-item--selected'),
    ).toEqual(false);

    fireEvent.click(secondTabLink);

    expect(
      firstTabLink?.parentElement?.classList.contains('nhsuk-tabs__list-item--selected'),
    ).toEqual(false);
    expect(
      secondTabLink?.parentElement?.classList.contains('nhsuk-tabs__list-item--selected'),
    ).toEqual(true);
  });

  // describe('The tabs title', () => {
  //   it.each`
  //     headingLevel
  //     ${undefined}
  //     ${'H1'}
  //     ${'H2'}
  //     ${'H3'}
  //     ${'H4'}
  //   `(
  //     'Renders the chosen heading level $headingLevel if specified',
  //     ({ headingLevel }: { headingLevel: HeadingLevelType }) => {
  //       const { container } = render(
  //         <Tabs.Title headingLevel={headingLevel}>Test title</Tabs.Title>,
  //       );

  //       const title = container.querySelector('.nhsuk-tabs__title');

  //       expect(title?.nodeName).toEqual(headingLevel ?? 'H2');
  //     },
  //   );
  // });

  describe('The tab list', () => {
    it('Renders the expected children', () => {
      const { container } = render(
        <Tabs.List>
          <div id="list-contents" />
        </Tabs.List>,
      );

      const listElement = container.querySelector('.nhsuk-tabs__list');

      expect(listElement?.querySelector('#list-contents')).toBeTruthy();
    });
  });

  describe('The tab list item', () => {
    it('Sets the href to be the passed in id prop', () => {
      const { container } = render(
        <Tabs.ListItem id="test-id">
          <div id="list-item-contents" />
        </Tabs.ListItem>,
      );

      expect(container.querySelector('.nhsuk-tabs__tab')?.getAttribute('href')).toBe('#test-id');
    });

    it('Renders the expected children', () => {
      const { container } = render(
        <Tabs.ListItem id="test-id">
          <div id="list-item-contents" />
        </Tabs.ListItem>,
      );

      const tabElement = container.querySelector('.nhsuk-tabs__tab');

      expect(tabElement?.querySelector('#list-item-contents')).toBeTruthy();
    });
  });

  describe('The tab contents', () => {
    it('Renders the expected children', () => {
      const { container } = render(
        <Tabs.Contents id="test-contents">
          <div id="tab-contents" />
        </Tabs.Contents>,
      );

      const tabElement = container.querySelector('#test-contents');

      expect(tabElement?.querySelector('#tab-contents')).toBeTruthy();
    });
  });
});
