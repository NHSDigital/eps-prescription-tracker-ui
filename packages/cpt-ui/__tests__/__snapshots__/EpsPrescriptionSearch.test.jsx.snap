// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The tabs component Matches the snapshot 1`] = `
<div>
  <div
    class="nhsuk-tabs"
    data-module="nhsuk-tabs"
  >
    <h2
      class="nhsuk-tabs__title"
    >
      Contents
    </h2>
    <ul
      class="nhsuk-tabs__list"
      role="tablist"
    >
      <li
        class="nhsuk-tabs__list-item nhsuk-tabs__list-item--selected"
        role="presentation"
      >
        <a
          aria-controls="IdSearch"
          aria-selected="true"
          class="nhsuk-tabs__tab"
          href="#IdSearch"
          id="tab_IdSearch"
          role="tab"
          tabindex="0"
        >
          Prescription ID search
        </a>
      </li>
      <li
        class="nhsuk-tabs__list-item"
        role="presentation"
      >
        <a
          aria-controls="NhsNumSearch"
          aria-selected="false"
          class="nhsuk-tabs__tab"
          href="#NhsNumSearch"
          id="tab_NhsNumSearch"
          role="tab"
          tabindex="-1"
        >
          NHS number search
        </a>
      </li>
      <li
        class="nhsuk-tabs__list-item"
        role="presentation"
      >
        <a
          aria-controls="BasicDetailsSearch"
          aria-selected="false"
          class="nhsuk-tabs__tab"
          href="#BasicDetailsSearch"
          id="tab_BasicDetailsSearch"
          role="tab"
          tabindex="-1"
        >
          Basic details search
        </a>
      </li>
    </ul>
    <div
      aria-labelledby="tab_IdSearch"
      class="nhsuk-tabs__panel"
      id="IdSearch"
      role="tabpanel"
    >
      <div>
        Prescription ID search
      </div>
    </div>
    <div
      aria-labelledby="tab_NhsNumSearch"
      class="nhsuk-tabs__panel nhsuk-tabs__panel--hidden"
      id="NhsNumSearch"
      role="tabpanel"
    >
      <div>
        NHS number search
      </div>
    </div>
    <div
      aria-labelledby="tab_BasicDetailsSearch"
      class="nhsuk-tabs__panel nhsuk-tabs__panel--hidden"
      id="BasicDetailsSearch"
      role="tabpanel"
    >
      <div>
        Basic details search
      </div>
    </div>
  </div>
</div>
`;
